src/sentry/tagstore/snuba/backend.py

```diff
@@ -173,33 +178,75 @@ def __get_tag_keys_for_projects(
         limit=1000,
         keys=None,
         include_values_seen=True,
+        use_cache=False,
         **kwargs
     ):
-        filters = {"project_id": projects}
+        default_start, default_end = default_start_end_dates()
+        if start is None:
+            start = default_start
+        if end is None:
+            end = default_end
+
+        filters = {"project_id": sorted(projects)}
         if environments:
-            filters["environment"] = environments
+            filters["environment"] = sorted(environments)
         if group_id is not None:
             filters["group_id"] = [group_id]
         if keys is not None:
-            filters["tags_key"] = keys
+            filters["tags_key"] = sorted(keys)
         aggregations = [["count()", "", "count"]]
 
         if include_values_seen:
             aggregations.append(["uniq", "tags_value", "values_seen"])
         conditions = []
 
-        result = snuba.query(
-            start=start,
-            end=end,
-            groupby=["tags_key"],
-            conditions=conditions,
-            filter_keys=filters,
-            aggregations=aggregations,
-            limit=limit,
-            orderby="-count",
-            referrer="tagstore.__get_tag_keys",
-            **kwargs
-        )
+        should_cache = use_cache and group_id is None
+        result = None
+
+        if should_cache:
+            filtering_strings = [
+                u"{}={}".format(key, value) for key, value in six.iteritems(filters)
+            ]
+            cache_key = u"tagstore.__get_tag_keys:{}".format(
+                md5_text(*filtering_strings).hexdigest()
+            )
+            key_hash = hash(cache_key)
+            should_cache = (key_hash % 1000) / 1000.0 <= options.get(
+                "snuba.tagstore.cache-tagkeys-rate"
+            )
+
+        # If we want to continue attempting to cache after checking against the cache rate
+        if should_cache:
+            # Round times by 5 minutes since we cache for 5 minutes
+            start = start.replace(
+                minute=start.minute // 5 * 5 + key_hash % 5, second=key_hash % 60, microsecond=0
+            )
+            end = end.replace(
+                minute=end.minute // 5 * 5 + key_hash % 5, second=key_hash % 60, microsecond=0
+            )
+            cache_key += u":{}:{}".format(start.isoformat(), end.isoformat())
```

Comment:
Also the jitter is not doing what is meant to do here.
Example:
it is 10:13. start:5:13 end:10:13
let's assume the jitter calculated adds 2 for this cache key. The resulting start and end will be
start: 5:12 (5:10 + 2) end: 10:12
your cache key is tagstore......HASH:5:12:10:12 .  (you should not use : as a separator probably)

Now we run the query for the same project at 10:15.
WIth the same jitter the calculated cache key would be 
tagstore......HASH:5:17:10:17 .  (15 + 2)
So the cache will be refreshed at 10:15, not at 10:17 thus creating the load spike you are trying to prevent with the jitter because everyone will rotate the cache key at the same time  no matter what the cache key actually is.

(the unit test is quite important to catch these issues)